zetaclient/chains/bitcoin/signer/sign.go

```diff
@@ -0,0 +1,239 @@
+	"github.com/btcsuite/btcd/wire"
+	"github.com/pkg/errors"
+
+	"github.com/zeta-chain/node/pkg/chains"
+	"github.com/zeta-chain/node/pkg/constant"
+	"github.com/zeta-chain/node/zetaclient/chains/bitcoin/common"
+	"github.com/zeta-chain/node/zetaclient/chains/bitcoin/observer"
+)
+
+const (
+	// the maximum number of inputs per outbound
+	MaxNoOfInputsPerTx = 20
+
+	// the rank below (or equal to) which we consolidate UTXOs
+	consolidationRank = 10
+)
+
+// SignWithdrawTx signs a BTC withdrawal tx and returns the signed tx
+func (signer *Signer) SignWithdrawTx(
+	ctx context.Context,
+	txData *OutboundData,
+	ob *observer.Observer,
+) (*wire.MsgTx, error) {
+	nonceMark := chains.NonceMarkAmount(txData.nonce)
+	estimateFee := float64(txData.feeRate*common.OutboundBytesMax) / 1e8
+	totalAmount := txData.amount + estimateFee + float64(nonceMark)*1e-8
+
+	// refresh unspent UTXOs and continue with keysign regardless of error
+	if err := ob.FetchUTXOs(ctx); err != nil {
+		signer.Logger().Std.Error().Err(err).Uint64("nonce", txData.nonce).Msg("SignWithdrawTx: FetchUTXOs failed")
+	}
+
+	// select N UTXOs to cover the total expense
+	selected, err := ob.SelectUTXOs(
+		ctx,
+		totalAmount,
+		MaxNoOfInputsPerTx,
+		txData.nonce,
+		consolidationRank,
+		false,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// build tx and add inputs
+	tx := wire.NewMsgTx(wire.TxVersion)
+	inAmounts, err := signer.AddTxInputs(tx, selected.UTXOs)
+	if err != nil {
+		return nil, err
+	}
+
+	// size checking
+	// #nosec G115 always positive
+	txSize, err := common.EstimateOutboundSize(int64(len(selected.UTXOs)), []btcutil.Address{txData.to})
+	if err != nil {
+		return nil, err
+	}
+	if txData.txSize < common.BtcOutboundBytesWithdrawer { // ZRC20 'withdraw' charged less fee from end user
+		signer.Logger().Std.Info().
+			Msgf("txSize %d is less than BtcOutboundBytesWithdrawer %d for nonce %d", txData.txSize, txSize, txData.nonce)
+	}
+	if txSize < common.OutboundBytesMin { // outbound shouldn't be blocked by low sizeLimit
+		signer.Logger().Std.Warn().
+			Msgf("txSize %d is less than outboundBytesMin %d; use outboundBytesMin", txSize, common.OutboundBytesMin)
+		txSize = common.OutboundBytesMin
+	}
+	if txSize > common.OutboundBytesMax { // in case of accident
+		signer.Logger().Std.Warn().
+			Msgf("txSize %d is greater than outboundBytesMax %d; use outboundBytesMax", txSize, common.OutboundBytesMax)
+		txSize = common.OutboundBytesMax
+	}
+
+	// fee calculation
+	// #nosec G115 always in range (checked above)
+	fees := txSize * txData.feeRate
+	signer.Logger().
+		Std.Info().
+		Msgf("bitcoin outbound nonce %d feeRate %d size %d fees %d consolidated %d utxos of value %v",
+			txData.nonce, txData.feeRate, txSize, fees, selected.ConsolidatedUTXOs, selected.ConsolidatedValue)
+
+	// add tx outputs
+	inputValue := selected.Value
+	err = signer.AddWithdrawTxOutputs(tx, txData.to, inputValue, txData.amountSats, nonceMark, fees, txData.cancelTx)
+	if err != nil {
```

Comment:
NIT but we be cleaner to use
```
if err := .. ; err != {
}
```
whenever possible

Reply:
updated https://github.com/zeta-chain/node/pull/3381/commits/3f5087b2b7a10b24e7ce5831dd65fcdf0f385216